soroban-env-host/src/budget.rs

```diff
@@ -111,33 +33,75 @@ impl CostType {
 /// The parameters for a `CostModel` are calibrated empirically. See this crate's
 /// benchmarks for more details.
 
-#[derive(Default, Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
-pub struct CostModel {
-    pub(crate) const_param: u64,
-    pub(crate) lin_param: u64,
+// #[derive(Default, Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+// pub struct CostModel {
+//     pub(crate) const_param: u64,
+//     pub(crate) lin_param: u64,
+// }
+
+// impl CostModel {
+//     pub fn from_xdr(xdr: &ContractCostParamEntry) -> Self {
+//         CostModel {
+//             const_param: xdr.const_term as u64,
+//             lin_param: xdr.linear_term as u64,
+//         }
+//     }
+
+//     /// Evaluates the linear model `f(input) = const_param + lin_param * Option<input>`.
+//     /// If the input is `None`, then the linear part is ingored, resulting in a constant
+//     /// model.
+//     pub fn evaluate(&self, input: Option<u64>) -> u64 {
+//         match input {
+//             Some(input) => {
+//                 let mut res = self.const_param;
+//                 if self.lin_param != 0 {
+//                     res = res.saturating_add(self.lin_param.saturating_mul(input));
+//                 }
+//                 res
+//             }
+//             None => self.const_param,
+//         }
+//     }
+
+//     #[cfg(test)]
+//     pub fn reset(&mut self) {
+//         self.const_param = 0;
+//         self.lin_param = 0;
+//     }
+// }
+
+pub trait HostCostModel {
+    fn evaluate(&self, input: Option<u64>) -> Result<u64, HostError>;
+
+    #[cfg(test)]
+    fn reset(&mut self);
 }
 
-impl CostModel {
-    /// Evaluates the linear model `f(input) = const_param + lin_param * Option<input>`.
-    /// If the input is `None`, then the linear part is ingored, resulting in a constant
-    /// model.
-    pub fn evaluate(&self, input: Option<u64>) -> u64 {
+impl HostCostModel for ContractCostParamEntry {
+    fn evaluate(&self, input: Option<u64>) -> Result<u64, HostError> {
+        if self.const_term < 0 || self.linear_term < 0 {
+            // TODO: consider more concrete error code "malformed xdr input"
+            return Err(ScUnknownErrorCode::Xdr.into());
+        }
+
+        let const_term = i64::from(self.const_term) as u64;
```

Comment:
Did we figure out whether the cost model types can be stored as u32?

Reply:
Last time we discussed why some the other network configuration entries have a mixture of unsigned and signed numbers. 

The `i32` type for the cost formula terms were chosen because of generality consideration: 1. if for some rare reason we may encounter future cost types that decrease with the declared input.  2. we may add higher order (beyond linear) terms (via `ext`) and negative parameters may become necessary. Both of these cases are far-fetched for now but wouldn't hurt to preserve that possibility. 
`i32` instead of `i64` was chosen because it should have enough digits to represent the largest single cost. Although it isn't out of the question we might need a larger number (a single const term with i32::MAX representing 2B instructions or 2GB memory) so we may want to consider i64.

Reply:
*shrug* ok